File: lib/gpu/lib/src/render_pass.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: public_member_api_docs

part of flutter_gpu;

base class ColorAttachment {
  ColorAttachment(
      {this.loadAction = LoadAction.clear,
      this.storeAction = StoreAction.store,
      this.clearValue = const ui.Color(0x00000000),
      required this.texture,
      this.resolveTexture = null});

  LoadAction loadAction;
  StoreAction storeAction;
  ui.Color clearValue;
  Texture texture;
  Texture? resolveTexture;
}

base class DepthStencilAttachment {
  DepthStencilAttachment(
      {this.depthLoadAction = LoadAction.clear,
      this.depthStoreAction = StoreAction.dontCare,
      this.depthClearValue = 0,
      this.stencilLoadAction = LoadAction.clear,
      this.stencilStoreAction = StoreAction.dontCare,
      this.stencilClearValue = 0,
      required this.texture});
  LoadAction depthLoadAction;
  StoreAction depthStoreAction;
  double depthClearValue;

  LoadAction stencilLoadAction;
  StoreAction stencilStoreAction;
  int stencilClearValue;

  Texture texture;
}

base class RenderPass extends NativeFieldWrapperClass1 {
  /// Creates a new RenderPass.
  RenderPass._(CommandBuffer commandBuffer, ColorAttachment colorAttachment,
      DepthStencilAttachment? depthStencilAttachment) {
    _initialize();
    String? error;
    error = _setColorAttachment(
        colorAttachment.loadAction.index,
        colorAttachment.storeAction.index,
        colorAttachment.clearValue.value,
        colorAttachment.texture,
        colorAttachment.resolveTexture);
    if (error != null) {
      throw Exception(error);
    }
    if (depthStencilAttachment != null) {
      error = _setDepthStencilAttachment(
          depthStencilAttachment.depthLoadAction.index,
          depthStencilAttachment.depthStoreAction.index,
          depthStencilAttachment.depthClearValue,
          depthStencilAttachment.stencilLoadAction.index,
          depthStencilAttachment.stencilStoreAction.index,
          depthStencilAttachment.stencilClearValue,
          depthStencilAttachment.texture);
      if (error != null) {
        throw Exception(error);
      }
    }
    error = _begin(commandBuffer);
    if (error != null) {
      throw Exception(error);
    }
  }

  void bindPipeline(RenderPipeline pipeline) {
    _bindPipeline(pipeline);
  }

  void bindVertexBuffer(BufferView bufferView, int vertexCount) {
    bufferView.buffer._bindAsVertexBuffer(
        this, bufferView.offsetInBytes, bufferView.lengthInBytes, vertexCount);
  }

  void bindUniform(UniformSlot slot, BufferView bufferView) {
    bool success = bufferView.buffer._bindAsUniform(
        this, slot, bufferView.offsetInBytes, bufferView.lengthInBytes);
    if (!success) {
      throw Exception("Failed to bind uniform slot");
    }
  }

  void draw() {
    if (!_draw()) {
      throw Exception("Failed to append draw");
    }
  }

  /// Wrap with native counterpart.
  @Native<Void Function(Handle)>(
      symbol: 'InternalFlutterGpu_RenderPass_Initialize')
  external void _initialize();

  @Native<Handle Function(Pointer<Void>, Int, Int, Int, Pointer<Void>, Handle)>(
      symbol: 'InternalFlutterGpu_RenderPass_SetColorAttachment')
  external String? _setColorAttachment(int loadAction, int storeAction,
      int clearColor, Texture texture, Texture? resolveTexture);

  @Native<
          Handle Function(
              Pointer<Void>, Int, Int, Float, Int, Int, Int, Pointer<Void>)>(
      symbol: 'InternalFlutterGpu_RenderPass_SetDepthStencilAttachment')
  external String? _setDepthStencilAttachment(
      int depthLoadAction,
      int depthStoreAction,
      double depthClearValue,
      int stencilLoadAction,
      int stencilStoreAction,
      int stencilClearValue,
      Texture texture);

  @Native<Handle Function(Pointer<Void>, Pointer<Void>)>(
      symbol: 'InternalFlutterGpu_RenderPass_Begin')
  external String? _begin(CommandBuffer commandBuffer);

  @Native<Void Function(Pointer<Void>, Pointer<Void>)>(
      symbol: 'InternalFlutterGpu_RenderPass_BindPipeline')
  external void _bindPipeline(RenderPipeline pipeline);

  @Native<Void Function(Pointer<Void>, Pointer<Void>, Int, Int, Int)>(
      symbol: 'InternalFlutterGpu_RenderPass_BindVertexBufferDevice')
  external void _bindVertexBufferDevice(DeviceBuffer buffer, int offsetInBytes,
      int lengthInBytes, int vertexCount);

  @Native<Void Function(Pointer<Void>, Pointer<Void>, Int, Int, Int)>(
      symbol: 'InternalFlutterGpu_RenderPass_BindVertexBufferHost')
  external void _bindVertexBufferHost(
      HostBuffer buffer, int offsetInBytes, int lengthInBytes, int vertexCount);

  @Native<Bool Function(Pointer<Void>, Int, Int, Pointer<Void>, Int, Int)>(
      symbol: 'InternalFlutterGpu_RenderPass_BindUniformDevice')
  external bool _bindUniformDevice(int stage, int slotId, DeviceBuffer buffer,
      int offsetInBytes, int lengthInBytes);

  @Native<Bool Function(Pointer<Void>, Int, Int, Pointer<Void>, Int, Int)>(
      symbol: 'InternalFlutterGpu_RenderPass_BindUniformHost')
  external bool _bindUniformHost(int stage, int slotId, HostBuffer buffer,
      int offsetInBytes, int lengthInBytes);

  @Native<Bool Function(Pointer<Void>)>(
      symbol: 'InternalFlutterGpu_RenderPass_Draw')
  external bool _draw();
}

File: lib/gpu/render_pass.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/gpu/render_pass.h"

#include "dart_api.h"
#include "impeller/core/formats.h"
#include "impeller/geometry/color.h"
#include "tonic/converter/dart_converter.h"

namespace flutter {
namespace gpu {

IMPLEMENT_WRAPPERTYPEINFO(flutter_gpu, RenderPass);

RenderPass::RenderPass() = default;

impeller::Command& RenderPass::GetCommand() {
  return command_;
}

impeller::RenderTarget& RenderPass::GetRenderTarget() {
  return render_target_;
}

bool RenderPass::Begin(flutter::gpu::CommandBuffer& command_buffer) {
  render_pass_ =
      command_buffer.GetCommandBuffer()->CreateRenderPass(render_target_);
  if (!render_pass_) {
    return false;
  }
  command_buffer.AddRenderPass(render_pass_);
  return true;
}

RenderPass::~RenderPass() = default;

}  // namespace gpu
}  // namespace flutter

static impeller::Color ToImpellerColor(uint32_t argb) {
  return impeller::Color::MakeRGBA8((argb >> 16) & 0xFF,  // R
                                    (argb >> 8) & 0xFF,   // G
                                    argb & 0xFF,          // B
                                    argb >> 24);          // A
}

//----------------------------------------------------------------------------
/// Exports
///

void InternalFlutterGpu_RenderPass_Initialize(Dart_Handle wrapper) {
  auto res = fml::MakeRefCounted<flutter::gpu::RenderPass>();
  res->AssociateWithDartWrapper(wrapper);
}

Dart_Handle InternalFlutterGpu_RenderPass_SetColorAttachment(
    flutter::gpu::RenderPass* wrapper,
    int load_action,
    int store_action,
    int clear_color,
    flutter::gpu::Texture* texture,
    Dart_Handle resolve_texture_wrapper) {
  impeller::ColorAttachment desc;
  desc.load_action = static_cast<impeller::LoadAction>(load_action);
  desc.store_action = static_cast<impeller::StoreAction>(store_action);
  desc.clear_color = ToImpellerColor(static_cast<uint32_t>(clear_color));
  desc.texture = texture->GetTexture();
  if (!Dart_IsNull(resolve_texture_wrapper)) {
    flutter::gpu::Texture* resolve_texture =
        tonic::DartConverter<flutter::gpu::Texture*>::FromDart(
            resolve_texture_wrapper);
    desc.resolve_texture = resolve_texture->GetTexture();
  }
  wrapper->GetRenderTarget().SetColorAttachment(desc, 0);
  return Dart_Null();
}

Dart_Handle InternalFlutterGpu_RenderPass_SetStencilAttachment(
    flutter::gpu::RenderPass* wrapper,
    int load_action,
    int store_action,
    int clear_stencil,
    flutter::gpu::Texture* texture) {
  impeller::StencilAttachment desc;
  desc.load_action = static_cast<impeller::LoadAction>(load_action);
  desc.store_action = static_cast<impeller::StoreAction>(store_action);
  desc.clear_stencil = clear_stencil;
  desc.texture = texture->GetTexture();
  wrapper->GetRenderTarget().SetStencilAttachment(desc);
  return Dart_Null();
}

Dart_Handle InternalFlutterGpu_RenderPass_Begin(
    flutter::gpu::RenderPass* wrapper,
    flutter::gpu::CommandBuffer* command_buffer) {
  if (!wrapper->Begin(*command_buffer)) {
    return tonic::ToDart("Failed to begin RenderPass");
  }
  return Dart_Null();
}
